### PR TITLE
Add cli to submit gov proposal to upgrade multiple channels

### PR DESCRIPTION
**Pull request by chatton**
## Description

closes: #5540

This PR adds a CLI to which queries all channels, and allows for specifying a pattern for the portID, and an optional comma separated list for channelIDs which will either generate a `proposal.json` file which contains `MsgChanUpgradeInit`s, or actually broadcast a `MsgSubmitProposal` directly.

`--dry-run` will behave the same way as `--json`.

I can add documentation in a follow up if this approach gets merged.

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against the correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/build/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

### CODE REVIEW

File: modules/core/04-channel/client/cli/tx.go [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+
+			for _, ch := range resp.Channels {
+
+				if !channelShouldBeUpgraded(*ch, pattern, channelIDs) {
+					continue
+				}
+
+				// construct a MsgChannelUpgradeInit which will upgrade the specified channel to a specific version.
+				msgUpgradeInit := types.NewMsgChannelUpgradeInit(ch.PortId, ch.ChannelId, types.NewUpgradeFields(ch.Ordering, ch.ConnectionHops, versionStr), clientCtx.GetFromAddress().String())
+				msgs = append(msgs, msgUpgradeInit)
```
Comment (chatton) [inline modules/core/04-channel/client/cli/tx.go:103]: I left ordering and connection hops unchanged, and only the version string changes. Technically we can allow these fields to be updated, but I think the version string will be all that we need 99% of the time.

File: modules/core/04-channel/client/cli/tx.go [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+	flagExpedited   = "expedited"
+	flagChannelIDs  = "channel-ids"
+)
+
+func newUpgradeChannelsCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "upgrade-channels",
+		Short: "Upgrade IBC channels",
+		Long: `Submit a governance proposal to upgrade all open channels whose port matches a specified pattern 
+(the default is transfer), optionally, specific an exact list of channel IDs with a comma separated list.`,
```
Comment (colin-axner) [inline modules/core/04-channel/client/cli/tx.go:36]:
```suggestion
(the default is transfer), optionally, an exact list of comma separated channel IDs may be specified.`,
```

**Review by colin-axner — APPROVED** [review]
Thank you @chatton! Looks great to me!

File: modules/core/04-channel/client/cli/tx.go (hunk 2 above)
Comment (charleenfei) [reply modules/core/04-channel/client/cli/tx.go:36]: should we specify in this docstring that if channel IDs are specified then it must match the specified port? or leave that for the longer documentation?

**Review by charleenfei — APPROVED** [review]
just one doc related comment, thanks @chatton!

File: modules/core/04-channel/client/cli/tx.go [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+	cmd.Flags().String(flagMetadata, "", "Metadata for the gov proposal which will upgrade existing open channels.")
+	cmd.Flags().String(flagPortPattern, "transfer", "The pattern to use to match port ids.")
+	cmd.Flags().Bool(flagExpedited, false, "set the expedited value for the governance proposal.")
+	cmd.Flags().Bool(flagJSON, false, "specify true to output valid proposal.json contents, instead of submitting a governance proposal.")
+	cmd.Flags().String(flagChannelIDs, "", "a comma separated list of channel IDs to upgrade.")
+
+	return cmd
+}
+
+// channelShouldBeUpgraded returns a boolean indicated whether or not the given channel should be upgraded based
```
Comment (crodriguezvega) [inline modules/core/04-channel/client/cli/tx.go:145]:
```suggestion
// channelShouldBeUpgraded returns a boolean indicating whether or not the given channel should be upgraded based
```

**Review by crodriguezvega — APPROVED** [review]
Thank you for the quick turnaround, @chatton! I just left a suggestion to maybe simplify some code.

File: modules/core/04-channel/client/cli/tx.go [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+
+			displayJSON, err := cmd.Flags().GetBool(flagJSON)
+			if err != nil {
+				return err
+			}
+
+			expidited, err := cmd.Flags().GetBool(flagExpedited)
+			if err != nil {
+				return err
+			}
```
[the remark this answers is not in the document]
Comment (DimitrisJim) [reply modules/core/04-channel/client/cli/tx.go:76]: yea, could also add `AddGovPropFlagsToCmd` though we do set a sensible default here so fine either way (maybe add and override default value?)

[the remark this answers is not in the document]
Comment (chatton) [reply modules/core/04-channel/client/cli/tx.go:76]: great suggestions, will implement these.

**Comment by chatton** [comment]
> just one doc related comment, thanks @chatton!

I think we can leave this for the longer docs

File: modules/core/04-channel/client/cli/cli.go [hunk 5]
```diff
@@ -50,7 +50,8 @@ func NewTxCmd() *cobra.Command {
 	}
 
 	txCmd.AddCommand(
-		NewPruneAcknowledgementsTxCmd(),
+		newUpgradeChannelsTxCmd(),
```
Comment (chatton) [inline modules/core/04-channel/client/cli/cli.go:53]: unexported this to stay consistent, I don't think this needs to be exported.

**Pull request merged**